Cache git dependencies as wheels

maksbotan opened the pull request:
Currently, poetry install will clone, build and install every git dependency when it's not present in the environment. This is OK for developer's machines, but not OK for CI - there environment is always fresh, and installing git dependencies takes significant time on each CI run, especially if the dependency has C extensions that need to be built.

This commit builds a wheel for every git dependency that has precise reference hash in lock file and is not required to be in editable mode, stores that wheel in a cache dir and will install from it instead of cloning the repository again.

# Pull Request Check List

<!-- This is just a reminder about the most common mistakes. Please make sure that you tick all *appropriate* boxes.  But please read our [contribution guide](https://python-poetry.org/docs/contributing/) at least once, it will save you unnecessary review cycles! -->

- [ ] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

This is a copy of #6896, which got automatically closed when `feature/wheel-installer-and-builder` branch was merged and deleted.

@neersighted @radoering may I ping you about this feature? 🙏 

If it's OK, I'll try to write a test.

maksbotan commented:
@sdispater may I ask your attention about this? Since 1.4.0 was released, this change can be incorporated into master. IMO this is an important improvement.

radoering commented on src/poetry/installation/executor.py:
```diff
@@ -582,14 +582,25 @@ def _prepare_git_archive(self, operation: Install | Update) -> Path:
         from poetry.vcs.git import Git
 
         package = operation.package
+        reference = package.source_resolved_reference or package.source_reference
```
Can we be sure that reference is always a hash or tag and never a branch?

radoering replied:
Sounds good to me.

radoering commented on src/poetry/installation/executor.py:
```diff
@@ -582,14 +582,25 @@ def _prepare_git_archive(self, operation: Install | Update) -> Path:
+        assert package.source_url is not None
+        assert reference is not None
+
+        cache_dir = self._chef.get_cache_directory_for_git(
+            package.source_url, reference
+        )
+        cache_dir.mkdir(parents=True, exist_ok=True)
+        cached_archive = next(cache_dir.glob("*.whl"), None)
+        if cached_archive is not None and not operation.package.develop:
+            return cached_archive
```
We should extract this to method in `ArtifactCache` similar to `get_cached_archive_for_link`.

radoering commented on src/poetry/installation/executor.py:
```diff
@@ -604,7 +615,13 @@ def _prepare_git_archive(self, operation: Install | Update) -> Path:
+        if operation.package.develop:
+            return archive
+        else:
+            cached_archive = cache_dir / archive.name
+            shutil.copy(archive, cached_archive)
```
Can we move instead of copy? Alternatively, we could pass an optional destination to `_prepare_archive`.

maksbotan replied:
I ended up using the second suggestions, refactoring artifacts cache a bit, making git cache use the same logic to find correct archive as is done for ordinary packages. This is more safe then just grabbing the first wheel found in the cache directory.

radoering commented on src/poetry/installation/chef.py:
```diff
@@ -256,6 +256,12 @@ def get_cache_directory_for_link(self, link: Link) -> Path:
 
+        return self._get_directory_from_hash(key_parts)
+
+    def get_cache_directory_for_git(self, url: str, ref: str) -> Path:
```
`subdirectory` is relevant, too

maksbotan replied:
Thanks!

radoering reviewed (changes requested):
Please note #7621 and see my review comments.

Further, I wonder if it might make sense to mark the cache directories of git dependencies somehow for debugging purposes, e.g. by creating some `created_from_git_dependency` file? This would also give you an easy way to get rid of all cached wheels for git dependencies without nuking your complete artifacts cache. 🤔

maksbotan commented:
Thank you for the review, @radoering. I've addressed your comments in a separate commit, I'll squash everything before the merge.

> Further, I wonder if it might make sense to mark the cache directories of git dependencies somehow for debugging purposes, e.g. by creating some `created_from_git_dependency` file? This would also give you an easy way to get rid of all cached wheels for git dependencies without nuking your complete artifacts cache. 🤔

I think this would make sense. What is a proper way to touch a file in `poetry`, or should I just grab `shutil` or something for that?

maksbotan reviewed (commented):
Some tests fail after my refactoring, I'll look at that later, if the overall approach is ok.

radoering commented:
> What is a proper way to touch a file in `poetry`, or should I just grab `shutil` or something for that?

[path.touch()](https://docs.python.org/3/library/pathlib.html#pathlib.Path.touch)

maksbotan commented:
@radoering i've addressed your comments and fixed failing tests.

radoering commented on src/poetry/installation/executor.py:
```diff
@@ -603,34 +604,66 @@ def _prepare_archive(self, operation: Install | Update) -> Path:
+        assert package.source_url is not None
+        assert reference is not None
+
+        if package.source_resolved_reference is not None:
```
```suggestion
        if package.source_resolved_reference:
```

just in case it's an empty string

radoering commented on src/poetry/installation/executor.py:
```diff
@@ -603,34 +604,66 @@ def _prepare_archive(self, operation: Install | Update) -> Path:
             url=package.source_url,
             source_root=self._env.path / "src",
-            revision=package.source_resolved_reference or package.source_reference,
+            revision=reference,
```
We can revert this change (and storing reference in a variable some lines above) since we only use `source_resolved_reference` for caching now.

radoering commented on src/poetry/installation/executor.py:
```diff
@@ -603,34 +604,66 @@ def _prepare_archive(self, operation: Install | Update) -> Path:
-        archive = self._prepare_archive(operation)
+        output_dir = None
+        if (
+            package.source_resolved_reference is not None
```
```suggestion
            package.source_resolved_reference
```

radoering commented on src/poetry/utils/cache.py:
```diff
@@ -231,35 +231,68 @@ def get_cache_directory_for_link(self, link: Link) -> Path:
+        self, url: str, ref: str, subdirectory: str | None
+    ) -> Path:
+        return self._get_directory_from_hash(
+            {"url": url, "ref": ref, "subdirectory": subdirectory}
```
Since we add subdirectory only if it's present for links, I'd do the same here.

radoering commented on src/poetry/utils/cache.py:
```diff
@@ -289,6 +322,9 @@ def get_cached_archive_for_link(
     def _get_cached_archives_for_link(self, link: Link) -> list[Path]:
```
Not used anymore. Can be removed.

radoering reviewed (changes requested):
LGTM after some minor changes and with some tests.

- in `test_cache.py`, we should add tests for `get_cache_directory_for_git` and `get_cached_archive_for_git`
- in `test_executor.py`, we probably should extend `test_executor_should_write_pep610_url_references_for_git` similar to `test_executor_should_write_pep610_url_references_for_wheel_urls` or `test_executor_should_write_pep610_url_references_for_non_wheel_urls`

maksbotan reviewed (commented):
Hi @radoering!

I've addressed your comments and added some tests. I did not add a test for `get_cached_archive_for_git`, as it is a wrapper around `_get_cached_archive`, which is tested elsewhere anyway. Moreover, there was no test for `get_cached_archive_for_link` anyway.

radoering commented:
> Moreover, there was no test for `get_cached_archive_for_link` anyway.

Actually, there are two tests for `get_cached_archive_for_link`. 😉 Apart from that, you're right. I just added a small smoke test that checks that none of the internal assertions is raised.

Further, I fixed the lost cleanup for editable git installs. Thereby, I noticed that the content of PEP 610 `direct_url.json` had been wrong for editable git installs and fixed that, too. Please take a look at my commits and verify that they do not break anything for you.

maksbotan commented:
@radoering thank you! 

I've run a few tests with my production use case and this branch seems to work correctly.

Should I rebase these commits now? Or perhaps squash into one?

maksbotan commented:
Thank you! Could you give some estimate when this can hit a release? I understand that it would be in 1.5.x series?

radoering commented:
It will be in 1.5. I wouldn't expect it in April. 1.4.2 was released about a week ago and we didn't do too much for 1.5 until then.

maksbotan commented:
Thanks.
